### PR TITLE
reworking function loggers to enable tracking of user log volume

### PR DESCRIPTION
**Pull request by brettsam**
Addresses #3127 for v1.

In order for both `ILogger` and `TraceWriter` usages to be tracked, I added a new `UserLogMetricsTraceWriter` that is used only as a composite writer returned by the `FunctionTraceWriterFactory`.

We had some issues with our log layering which meant I had to do some refactoring. Namely, we assumed that the `TraceWriter` returned from the `FunctionTraceWriterFactory` was for file logging only, and filtered based on that. This meant that any logging of user log metrics from within this `TraceWriter` could be ignored if file logging was disabled. I moved this filtering around a bit to make this work and added new tests.

Also renamed the "wrapper" `FunctionLogger` to `FunctionLogHelper` to remove confusion and to sound less like an `ILogger`.

### CODE REVIEW

**Comment by brettsam** [comment]
@mathewc / @fabiocav -- ping. I'd like to get this into the next v1 release.

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 1]
```diff
@@ -841,7 +840,12 @@ private void ConfigureLogging()
             // to the startup logger until we've read configuration settings and can create the real logger.
             // The "startup" logger is used in this class for startup related logs. The public logger is used
             // for all other logging after startup.
-            FunctionTraceWriterFactory = new FunctionTraceWriterFactory(ScriptConfig);
+            FunctionTraceWriterFactory = new FunctionTraceWriterFactory(ScriptConfig, t =>
+            {
+                // File logging is done conditionally.
+                return FileLoggingEnabled && (!(t.Properties?.ContainsKey(ScriptConstants.TracePropertyPrimaryHostKey) ?? false) || IsPrimary);
+            });
```
Comment (paulbatum) [inline src/WebJobs.Script/Host/ScriptHost.cs]: this logic is a little hard to follow, between the negation, the or, and the null coalesce. Can you introduce one or more two named variables?

**Comment by brettsam** [comment]
Going to close this out. We've got this in for v2 (#3606) and all of this refactoring likely isn't worth it at this point for v1.
